integration/app_integration_test.go

```diff
@@ -0,0 +1,430 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/lib"
+	"github.com/gravitational/teleport/lib/auth/testauthority"
+	"github.com/gravitational/teleport/lib/httplib/csrf"
+	"github.com/gravitational/teleport/lib/service"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/web"
+	"github.com/gravitational/teleport/lib/web/app"
+
+	"github.com/gravitational/trace"
+
+	"github.com/pborman/uuid"
+	"github.com/stretchr/testify/require"
+)
+
+// TestForward tests that requests within get forwarded to the target
+// application within a single cluster and through a trusted cluster.
+func TestForward(t *testing.T) {
+	// Create cluster, user, and credentials package.
+	pack := setup(t)
+
+	// Create user for tests.
+	pack.createUser(t)
+
+	// Create Web UI session.
+	webCookie, webToken := pack.createWebSession(t)
+
+	tests := []struct {
+		desc          string
+		inCookie      string
+		outStatusCode int
+		outMessage    string
+	}{
+		{
+			desc:          "root cluster, valid application session cookie, success",
+			inCookie:      pack.createAppSession(t, webCookie, webToken, pack.rootAppPublicAddr, pack.rootAppClusterName),
+			outStatusCode: http.StatusOK,
+			outMessage:    pack.rootMessage,
+		},
+		{
+			desc:          "leaf cluster, valid application session cookie, success",
+			inCookie:      pack.createAppSession(t, webCookie, webToken, pack.leafAppPublicAddr, pack.leafAppClusterName),
+			outStatusCode: http.StatusOK,
+			outMessage:    pack.leafMessage,
+		},
+		{
+			desc:          "invalid application session cookie, redirect to login",
+			inCookie:      "D25C463CD27861559CC6A0A6AE54818079809AA8731CB18037B4B37A80C4FC6C",
+			outStatusCode: http.StatusFound,
+			outMessage:    "",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
```

Comment:
I'd create a copy of the test case's value in case the tests are made to run in parallel:
```go
for _, tt := range tests {
	tt := tt
	t.Run(tt.desc, func(t *testing.T) {
	...
```